### PR TITLE
PFAlgo range based loops and TestBench removal

### PR DESCRIPTION
**Pull request by guitargeek**
Continuous integration is a blessing. I am convinced that by regular tiny PRs to the PFProducer/PFAlgo that are easy to review we can on a long term clear up the Particle Flow hairball.

I suggest here:
  * to replace most loops in PFAlgo with range based loops
  * to replace boost::shared_ptr with std::shared_ptr
  * to remove the PFAlgoTestBench* classes since they are probably not used any more and IMHO not well designed: they had rather special requirements of the class they are testing (vtable, protected member functions) which I consider not a good design. Testing should be independent of implementation. Also, removing these classes makes further refactoring easier because they don't have to be considered. @cbernet, what do you think about this? Are these test benches still needed?

Local matrix tests pass.

### CODE REVIEW

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
> Reco comparison results: 252 differences found in the comparisons

charged hadron isolation changed for electrons and photons.
@guitargeek  please explain the reason.

**Comment by guitargeek** [comment]
Hi Slava, thank you for pointing that out, but how do you know that? I usually go to "comparisons with baseline" and then check all the directories there for "fails". Can you explain me how you check the differences?

There should be no difference at all. I thought I kept this PR small enough such that the "bug-surface" is minimal, but it is possible that I made a mistake.

**Comment by slava77** [comment]
> Hi Slava, thank you for pointing that out, but how do you know that? I usually go to "comparisons with baseline" and then check all the directories there for "fails". Can you explain me how you check the differences?
> 
> There should be no difference at all. I thought I kept this PR small enough such that the "bug-surface" is minimal, but it is possible that I made a mistake. I should eventually figure out a quick way to validate the PFProducer outside GitHub...

in the summary.html look for "Comparison failed"

**Comment by guitargeek** [comment]
Thanks, that's much easier.

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoParticleFlow/PFProducer/src/PFAlgo.cc [hunk 1]
```diff
@@ -3123,10 +3099,10 @@ void PFAlgo::processBlock( const reco::PFBlockRef& blockref,
 				  reco::PFBlock::LINKTEST_ALL );
 	
 	bool isClosest = true;
-	for(IE ih = hcalElems.begin(); ih != hcalElems.end(); ++ih ) {
+    for(auto const& hcal : hcalElems) {
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/src/PFAlgo.cc]: indent

File: RecoParticleFlow/PFProducer/src/PFAlgo.cc [hunk 2]
```diff
@@ -3698,8 +3669,8 @@ PFAlgo::postCleaning() {
 	
 	// Check that it is  not already scheculed to be cleaned
 	bool skip = false;
-	for(unsigned j=0; j<pfCandidatesToBeRemoved.size(); ++j) {
-	  if ( i == pfCandidatesToBeRemoved[j] ) skip = true;
+    for(unsigned int toRemove : pfCandidatesToBeRemoved) {
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/src/PFAlgo.cc]: indent

File: RecoParticleFlow/PFProducer/src/PFAlgo.cc [hunk 3]
```diff
@@ -3916,26 +3883,24 @@ void PFAlgo::setElectronExtraRef(const edm::OrphanHandle<reco::PFCandidateElectr
...
-    if((*pfCandidates_)[ic].particleId()==PFCandidate::gamma && (*pfCandidates_)[ic].mva_nothing_gamma() > 0.99) {
-      if((*pfCandidates_)[ic].superClusterRef().isNonnull()) {
+    if(cand.particleId()==PFCandidate::gamma && cand.mva_nothing_gamma() > 0.99) {
+      if(cand.superClusterRef().isNonnull()) {
 	bool found = false;
-	for(unsigned pcextra=0;pcextra<sizePhExtra;++pcextra) {
-	  if((*pfCandidates_)[ic].superClusterRef() == pfPhotonExtra_[pcextra].superClusterRef()) {
+	for(unsigned pcextra=0;pcextra<pfPhotonExtra_.size();++pcextra) {
+	  if(cand.superClusterRef() == pfPhotonExtra_[pcextra].superClusterRef()) {
 	    reco::PFCandidatePhotonExtraRef theRef(ph_extrah,pcextra);
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/src/PFAlgo.cc]: range loop here as well (increment/define the index separately)

**Comment by guitargeek** [comment]
Hi Slava, that's good to know that you also think the indentation and whitespaces should be brought into order. I was hesitant to do this so far.

I did some little improvements in the loop bodies where I fixed the indentation, so I'd recommend you to look as the latest commit changes with `?w=1` to highlight these non-whitespace changes.

https://github.com/cms-sw/cmssw/pull/25432/commits/8b3aa80fc3367f38b9ca62b9b59b829fe6cdbb1e?w=1

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoParticleFlow/PFProducer/src/PFAlgo.cc [hunk 4]
```diff
@@ -1451,127 +1435,114 @@ void PFAlgo::processBlock( const reco::PFBlockRef& blockref,
...
+        // Since the electrons were found beforehand, this track must be a hadron. Calibrate 
+        // the energy under the hadron hypothesis.
+        totalEcal += ecalEnergy;
+        const double previousCalibEcal = calibEcal;
+        const double previousSlopeEcal = slopeEcal;
+        calibEcal = std::max(totalEcal,0.);
+        calibHcal = 0.;
+        calibration_->energyEmHad(trackMomentum,calibEcal,calibHcal,
+                      clusterRef->positionREP().Eta(),
+                      clusterRef->positionREP().Phi());
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/src/PFAlgo.cc]: was it your editors decision to not align/indent to the opening brace?
Here it looks still appropriate to do it.

File: RecoParticleFlow/PFProducer/src/PFAlgo.cc [hunk 5]
```diff
@@ -1451,127 +1435,114 @@ void PFAlgo::processBlock( const reco::PFBlockRef& blockref,
...
+          // Turn this last cluster in a photon 
+          // (The PS clusters are already locked in "associatePSClusters")
+          active[index] = false;
+
+          // Find the associated tracks
+          std::multimap<double, unsigned> assTracks;
+          block.associatedElements( index,  linkData,
+                        assTracks,
+                        reco::PFBlockElement::TRACK,
+                        reco::PFBlock::LINKTEST_ALL );
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/src/PFAlgo.cc]: was it your editors decision to not align/indent to the opening brace?
Here it looks still appropriate to do it.

File: RecoParticleFlow/PFProducer/src/PFAlgo.cc [hunk 6]
```diff
@@ -2281,22 +2249,22 @@ void PFAlgo::processBlock( const reco::PFBlockRef& blockref,
...
       // the calorimetric energy does not saturate the total momentum.
-      if ( is->first < 0. || caloEnergy - totalChargedMomentum <= 0. ) { 
-	if(debug_) cout<<"\t\t\tactive, adding "<<is->second.second
+      if ( ecalSatellite.first < 0. || caloEnergy - totalChargedMomentum <= 0. ) { 
+	if(debug_) cout<<"\t\t\tactive, adding "<<ecalSatellite.second.second
 		       <<" to ECAL energy, and locking"<<endl;
-	active[is->second.first] = false;
-        double clusterEnergy=sqrt(is->second.second.Mag2());
+	active[ecalSatellite.second.first] = false;
+        double clusterEnergy=sqrt(ecalSatellite.second.second.Mag2());
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/src/PFAlgo.cc]: indentation

File: RecoParticleFlow/PFProducer/src/PFAlgo.cc [hunk 7]
```diff
@@ -2339,14 +2319,14 @@ void PFAlgo::processBlock( const reco::PFBlockRef& blockref,
       */
       // First consider loose muons
       if ( nMuons > 0 ) { 
-	for ( IT it = associatedTracks.begin(); it != associatedTracks.end(); ++it ) { 
-	  unsigned iTrack = it->second.first;
+    for(auto const& trk : associatedTracks) {
```
[the remark this answers is not in the document]
Comment (slava77) [reply RecoParticleFlow/PFProducer/src/PFAlgo.cc]: this didn't get fixed

File: RecoParticleFlow/PFProducer/src/PFAlgo.cc [hunk 8]
```diff
@@ -2860,16 +2840,15 @@ void PFAlgo::processBlock( const reco::PFBlockRef& blockref,
 	//       (*pfCandidates_)[tmpi].addElement(&elements[iPivotal]);
 	// (*pfCandidates_)[tmpi].addElementInBlock(blockref, iPivotal[iPivot]);
 	(*pfCandidates_)[tmpi].addElementInBlock( blockref, iHcal );
-	for ( unsigned ich=0; ich<chargedHadronsInBlock.size(); ++ich) { 
-	  unsigned iTrack = chargedHadronsInBlock[ich];
+    for (unsigned iTrack : chargedHadronsInBlock) {
```
[the remark this answers is not in the document]
Comment (slava77) [reply RecoParticleFlow/PFProducer/src/PFAlgo.cc]: here as well, still off

File: RecoParticleFlow/PFProducer/src/PFAlgo.cc (hunk 1 above)
Comment (slava77) [reply RecoParticleFlow/PFProducer/src/PFAlgo.cc]: and here

**Comment by cbernet** [comment]
@guitargeek actually I don't remember the PFAlgoTestBench classes, so as far as I'm concerned they can be removed (but others may know better).

File: RecoParticleFlow/PFProducer/src/PFAlgo.cc (hunk 4 above)
Comment (guitargeek) [reply RecoParticleFlow/PFProducer/src/PFAlgo.cc]: Apparently my vims decision was to indent spaces such that it looks like it was done with tabs before, yes. I corrected that.

**Comment by guitargeek** [comment]
Hi @slava77, I fixed the indentation issues you pointed out and did again minor changes in the blocks I reindented. Realizing how much of the code was already touched here by fixing the indentation, I then though it might also be the time to remove trailing whitespaces.

**Comment by slava77** [comment]
@cmsbuild please test

@guitargeek 
thank you for the update

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/25432	78f0f64
- code changes are in line with the PR description and the follow up review; all changes are technical and should not be affecting physics results
    - apparently at least some of the code re-indents were done manually (not automatically by an editor tool). So, it is not worth the effort to iterate on this.  RecoParticleFlow/PFProducer/src/PFAlgo.cc looks much better anyways.
- jenkins tests pass and comparisons with the baseline show no (relevant) differences

**Comment by guitargeek** [comment]
Hi Slava, thank you for insisting not further on the indentation, I know there are still some spots where I didn't get it right! There are more radical refactorings to come so the indentation levels will change again anyway I guess.

**Comment by fabiocos** [comment]
+1

**Pull request merged**